File: topic1/problem16/Solution.java
class Solution {
    public int trap(int[] height) {
        int area = 0;
        int l = 0, r = height.length-1;
        int lmax = height[l], rmax = height[r];

        while (l < r) {
            if (lmax < rmax) {
                ++l;
                lmax = Math.max(lmax, height[l]);
                area += lmax - height[l];
            } else {
                --r;
                rmax = Math.max(rmax, height[r]);
                area += rmax - height[r];
            }
        }
        return area;
    }
}

File: topic1/problem16/Solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        area = lmax = 0

        for i in range(1, len(height)):
            x = i

            while lmax < x and height[x] > height[x-1]:
                area += min(height[x], height[lmax]) - height[x-1]
                height[x-1] = min(height[x], height[lmax])
                x -= 1
            
            if height[lmax] < height[i]: lmax = i
        
        return area
